### PR TITLE
Log system refactor

### PR DESCRIPTION
**Pull request by alagroy-42**
Refactoring and testing the log classes.
- moving lizmapLogItem and lizmapLogConfig to Lizmap\Logger namespace.
- uncontextualizing Jelix from the classes
- Adding Unit tests for lizmapLogItem

### CODE REVIEW

File: lizmap/modules/lizmap/lib/App/JelixContext.php [hunk 1]
```diff
@@ -255,4 +267,12 @@ public function getFullUrl($selector, $params = array())
     {
         return \jUrl::getFull($selector, $params);
     }
+
+    public function getIniModifier($ini)
+    {
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/App/JelixContext.php:272]: I think the real content of this method is missing. there is no result property, no use of jIniModifier etc...

File: lizmap/modules/lizmap/classes/lizmapLogItem.class.php [hunk 2]
```diff
@@ -9,11 +9,12 @@
  *
  * @license Mozilla Public License : http://www.mozilla.org/MPL/
  */
+
+use Lizmap\App;
+use Lizmap\Logger as Log;
+
 class lizmapLogItem
```
Comment (laurentj) [inline lizmap/modules/lizmap/classes/lizmapLogItem.class.php:16]: mark  it as deprecated

File: lizmap/modules/lizmap/classes/lizmapLogConfig.class.php [hunk 3]
```diff
@@ -9,40 +9,24 @@
...
+use Lizmap\Logger as Log;
 
-    // list of the logItems of the ini file
-    protected $logItems = array();
-    // If the log is active globally or not
-    private $active = '';
-
-    // database profile
-    private $profile = '';
+class lizmapLogConfig
```
Comment (laurentj) [inline lizmap/modules/lizmap/classes/lizmapLogConfig.class.php:15]: mark it as deprecated. Probably we could remove it completely. We should inspect external modules to see if this class and lizmapLogItem are used directly.

File: lizmap/modules/lizmap/lib/Logger/Config.php [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    protected $logItems = array();
+    // If the log is active globally or not
+    private $active = '';
+
+    // database profile
+    private $profile = '';
+
+    protected $appContext;
+
+    public function __construct($readConfigPath, App\AppContextInterface $appContext)
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Logger/Config.php:38]: $readConfigPath is a wrong name. It does not content a path, right?

File: lizmap/modules/lizmap/lib/Logger/Config.php [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+    /**
+     * save the global configuration data.
+     *
+     * @param null|mixed $ini
+     */
+    public function save($ini = null)
+    {
+        if (!$ini) {
+            $iniFile = $this->appContext->appConfigPath('lizmapLogConfig.ini.php');
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Logger/Config.php:140]: I think it is better if we give the ini filename to the constructor, we could test it easily

File: lizmap/modules/lizmap/lib/Logger/Item.php [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+    /**
+     * Update the data for the log item in the ini file.
+     *
+     * @param mixed $data
+     *
+     * @return bool true if there were some modifications
+     */
+    public function update($data)
+    {
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Logger/Item.php:115]: what are these data? what data should contain? at least comments are missing. I would prefer some setter methods to set each data, and this method should only save $this->data. And i didn't find a usage of this method in the code. Probably it is not used at all.

**Review by laurentj — CHANGES_REQUESTED** [review]

**Comment by alagroy-42** [comment]
I couldn't find any references to the update/modify/save method in the LWC code, is it present in any module ? If not, should we delete it ?
@mdouchin

**Review by rldhont — APPROVED** [review]

**Review by laurentj — CHANGES_REQUESTED** [review]

File: lizmap/modules/lizmap/classes/lizmapLogConfig.class.php [hunk 7]
```diff
@@ -9,40 +9,27 @@
...
     {
-        $this->data = $readConfigPath;
-
-        // set generic parameters
-        foreach ($this->properties as $prop) {
-            if (isset($readConfigPath['general'][$prop])) {
-                $this->{$prop} = $readConfigPath['general'][$prop];
-            }
-        }
+        $this->config = new Log\Config($readConfigPath);
```
Comment (laurentj) [inline lizmap/modules/lizmap/classes/lizmapLogConfig.class.php:27]: app context parameter is missing

File: lizmap/modules/lizmap/classes/lizmapLogItem.class.php [hunk 8]
```diff
@@ -54,43 +38,37 @@ class lizmapLogItem
...
+        return $this->item->getKey();
     }
 
     /**
      * @return string[] list of properties name
      */
     public function getProperties()
     {
-        return self::$properties;
+        return $this->properties;
```
Comment (laurentj) [inline lizmap/modules/lizmap/classes/lizmapLogItem.class.php:71]: use Item::getSProperties instead

File: lizmap/modules/lizmap/classes/lizmapLogItem.class.php [hunk 9]
```diff
@@ -9,11 +9,15 @@
...
+/**
+ * @deprecated
+ */
 class lizmapLogItem
 {
-    // Lizmap log configuration file path (relative to the path folder)
-    private $config = 'config/lizmapLogConfig.ini.php';
-
     // log item properties
     private static $properties = array(
```
Comment (laurentj) [inline lizmap/modules/lizmap/classes/lizmapLogItem.class.php:22]: remove it and use static method of Item

File: lizmap/modules/lizmap/classes/lizmapLogItem.class.php [hunk 10]
```diff
@@ -23,26 +27,6 @@ class lizmapLogItem
...
-    // If user IP address must be logged in the detail log
-    private $logIp = '';
-
-    // If an email must be sent to the admin contact
-    private $logEmail = '';
-
-    private $data = array();
-
     // log record keys
     private static $recordKeys = array(
```
Comment (laurentj) [inline lizmap/modules/lizmap/classes/lizmapLogItem.class.php:31]: remove it

File: lizmap/modules/lizmap/classes/lizmapLogItem.class.php [hunk 11]
```diff
@@ -167,20 +119,7 @@ public function update($data)
...
-                $rec->{$k} = $data[$k];
-            }
-        }
-
-        try {
-            $dao->insert($rec);
-        } catch (Exception $e) {
-            jLog::log('Error while inserting a new line in log_detail :'.$e->getMessage());
-        }
+        return $this->item->insertLogDetail($data, $profile);
```
Comment (laurentj) [inline lizmap/modules/lizmap/classes/lizmapLogItem.class.php:122]: insertLogDetail does not return a value, so no return

File: tests/units/classes/Log/ConfigTest.php [hunk 12]
```diff
@@ -38,8 +52,8 @@ public function getTestModifyData()
 
     public function testModify($data, $newData, $expectedReturnValue)
     {
-        $testLizmapLogConfig = new lizmapLogConfig($data);
-        $this->assertEquals($expectedReturnValue, $testLizmapLogConfig->modify($newData));
+        $testLizmapLogConfig = new ConfigForTests($data, $this->context, null);
+        $this->assertEquals($expectedReturnValue, $testLizmapLogConfig->modifyForTests($newData));
         unset($testLizmapLogConfig);
```
Comment (laurentj) [inline tests/units/classes/Log/ConfigTest.php:57]: no need to unset

**Review by laurentj — CHANGES_REQUESTED** [review]

**Comment by laurentj** [comment]
Please fix lint issues showed by checks

File: lizmap/modules/lizmap/lib/Logger/Config.php [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+     *
+     * @return string[] list of names
+     */
+    public function getLogItemList()
+    {
+        $logItemList = array();
+
+        foreach ($this->data as $section => $data) {
+            if (strstr($section, 'item:')) {
+                $logItemList[] = preg_replace('#(^item:)#', '', $section, 1);
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Logger/Config.php:109]: Not optimized. Your code search two times in the string. 

Here is a better solution, search and retrieve at the same time:

```
           if (preg_match('/^item:(.*)/', $section, $matches)) {
                  $logItemList[] = $matches[1];
           }
```

**Review by laurentj — APPROVED** [review]

File: lizmap/modules/lizmap/lib/Logger/Config.php (hunk 13 above)
Comment (laurentj) [reply lizmap/modules/lizmap/lib/Logger/Config.php:109]: I'll fix it

**Pull request merged**